File: examples/canonical-faker/go/internal/validate/validate.go
package validate

import (
	"fmt"
	"strings"
	"time"

	"github.com/agent-context-base/canonical-faker-go/internal/domain"
	"github.com/agent-context-base/canonical-faker-go/internal/pools"
)

type ValidationReport struct {
	OK         bool           `json:"ok"`
	Violations []string       `json:"violations"`
	Counts     map[string]int `json:"counts"`
	Seed       int            `json:"seed"`
	Profile    string         `json:"profile"`
}

func Dataset(dataset domain.Dataset) ValidationReport {
	violations := []string{}
	counts := map[string]int{
		"organizations": len(dataset.Organizations),
		"users":         len(dataset.Users),
		"memberships":   len(dataset.Memberships),
		"projects":      len(dataset.Projects),
		"audit_events":  len(dataset.AuditEvents),
		"api_keys":      len(dataset.APIKeys),
		"invitations":   len(dataset.Invitations),
	}
	minimumRows := map[string]int{
		"organizations": 1,
		"users":         1,
		"memberships":   1,
		"projects":      1,
		"audit_events":  1,
		"api_keys":      0,
		"invitations":   0,
	}
	for entity, count := range counts {
		if count < minimumRows[entity] {
			violations = append(violations, fmt.Sprintf("row count below minimum for %s: %d < %d", entity, count, minimumRows[entity]))
		}
	}
	orgs := map[string]domain.Organization{}
	users := map[string]domain.User{}
	memberEmails := map[string]map[string]bool{}
	seenOrgIDs := map[string]bool{}
	seenEmails := map[string]bool{}
	seenPrefixes := map[string]bool{}

	for _, org := range dataset.Organizations {
		if seenOrgIDs[org.ID] {
			violations = append(violations, "duplicate organizations.id: "+org.ID)
		}
		seenOrgIDs[org.ID] = true
		orgs[org.ID] = org
	}
	for _, user := range dataset.Users {
		email := strings.ToLower(user.Email)
		if seenEmails[email] {
			violations = append(violations, "duplicate user.email: "+email)
		}
		seenEmails[email] = true
		users[user.ID] = user
	}
	graph := pools.Build(dataset)
	projects := map[string]domain.Project{}
	for _, project := range dataset.Projects {
		projects[project.ID] = project
	}
	membershipIDs := map[string]bool{}
	for _, membership := range dataset.Memberships {
		org, hasOrg := orgs[membership.OrgID]
		user, hasUser := users[membership.UserID]
		if !hasOrg {
			violations = append(violations, "membership missing org: "+membership.ID)
			continue
		}
		if !hasUser {
			violations = append(violations, "membership missing user: "+membership.ID)
			continue
		}
		membershipIDs[membership.ID] = true
		if memberEmails[membership.OrgID] == nil {
			memberEmails[membership.OrgID] = map[string]bool{}
		}
		memberEmails[membership.OrgID][strings.ToLower(user.Email)] = true
		if pools.ParseISO(membership.JoinedAt).Before(pools.ParseISO(org.CreatedAt)) {
			violations = append(violations, "Rule A violated by membership "+membership.ID)
		}
		if membership.InvitedBy != nil {
			if _, ok := users[*membership.InvitedBy]; !ok {
				violations = append(violations, "membership invited_by missing user: "+membership.ID)
			}
		}
	}
	apiKeyIDs := map[string]bool{}
	for _, key := range dataset.APIKeys {
		apiKeyIDs[key.ID] = true
		if !contains(graph.OrgMemberMap[key.OrgID], key.CreatedBy) {
			violations = append(violations, "Rule G violated by api_key "+key.ID)
		}
		if seenPrefixes[key.KeyPrefix] {
			violations = append(violations, "duplicate api_key.key_prefix: "+key.KeyPrefix)
		}
		seenPrefixes[key.KeyPrefix] = true
		if key.LastUsedAt != nil && pools.ParseISO(*key.LastUsedAt).Before(pools.ParseISO(key.CreatedAt)) {
			violations = append(violations, "api_key last_used_at before created_at: "+key.ID)
		}
	}
	invitationIDs := map[string]bool{}
	for _, invitation := range dataset.Invitations {
		invitationIDs[invitation.ID] = true
		if !contains(graph.OrgMemberMap[invitation.OrgID], invitation.InvitedBy) {
			violations = append(violations, "Rule H violated by invitation "+invitation.ID)
		}
		if memberEmails[invitation.OrgID][strings.ToLower(invitation.InvitedEmail)] {
			violations = append(violations, "Rule I violated by invitation "+invitation.ID)
		}
		expiresAt := pools.ParseISO(invitation.ExpiresAt)
		if !expiresAt.After(pools.BaseTime) {
			violations = append(violations, "invitation expiry must be in the future: "+invitation.ID)
		}
		if expiresAt.After(pools.BaseTime.Add(30 * 24 * time.Hour)) {
			violations = append(violations, "invitation expiry must be within 30 days: "+invitation.ID)
		}
		if invitation.AcceptedAt != nil && pools.ParseISO(*invitation.AcceptedAt).After(pools.BaseTime) {
			violations = append(violations, "invitation accepted_at must be in the past: "+invitation.ID)
		}
	}
	for _, project := range dataset.Projects {
		org, ok := orgs[project.OrgID]
		if !ok {
			violations = append(violations, "project missing org: "+project.ID)
			continue
		}
		if pools.ParseISO(project.CreatedAt).Before(pools.ParseISO(org.CreatedAt)) {
			violations = append(violations, "Rule B violated by project "+project.ID)
		}
		if !contains(graph.OrgMemberMap[project.OrgID], project.CreatedBy) {
			violations = append(violations, "Rule C violated by project "+project.ID)
		}
	}
	for _, event := range dataset.AuditEvents {
		project, ok := projects[event.ProjectID]
		if !ok {
			violations = append(violations, "audit event missing project: "+event.ID)
			continue
		}
		if _, hasOrg := orgs[event.OrgID]; !hasOrg {
			violations = append(violations, "audit event missing org: "+event.ID)
			continue
		}
		if !contains(graph.OrgMemberMap[event.OrgID], event.UserID) {
			violations = append(violations, "Rule D violated by audit event "+event.ID)
		}
		if project.OrgID != event.OrgID {
			violations = append(violations, "Rule E violated by audit event "+event.ID)
		}
		if pools.ParseISO(event.TS).Before(pools.ParseISO(project.CreatedAt)) {
			violations = append(violations, "Rule F violated by audit event "+event.ID)
		}
		membership := graph.MembershipByOrgUser[event.OrgID+":"+event.UserID]
		if !membershipIsZero(membership) && pools.ParseISO(event.TS).Before(pools.ParseISO(membership.JoinedAt)) {
			violations = append(violations, "audit event before membership joined_at: "+event.ID)
		}
		switch event.ResourceType {
		case "project":
			if _, ok := projects[event.ResourceID]; !ok {
				violations = append(violations, "audit event resource project missing: "+event.ID)
			}
		case "user":
			if _, ok := users[event.ResourceID]; !ok {
				violations = append(violations, "audit event resource user missing: "+event.ID)
			}
		case "membership":
			if !membershipIDs[event.ResourceID] {
				violations = append(violations, "audit event resource membership missing: "+event.ID)
			}
		case "api_key":
			if !apiKeyIDs[event.ResourceID] {
				violations = append(violations, "audit event resource api_key missing: "+event.ID)
			}
		case "invitation":
			if !invitationIDs[event.ResourceID] {
				violations = append(violations, "audit event resource invitation missing: "+event.ID)
			}
		}
	}

	return ValidationReport{
		OK:         len(violations) == 0,
		Violations: violations,
		Counts:     counts,
		Seed:       dataset.Seed,
		Profile:    dataset.ProfileName,
	}
}

func contains(values []string, target string) bool {
	for _, value := range values {
		if value == target {
			return true
		}
	}
	return false
}

func membershipIsZero(value domain.Membership) bool {
	return value.ID == ""
}

File: examples/canonical-faker/kotlin/src/main/kotlin/io/agentcontextbase/faker/validate/Validate.kt
package io.agentcontextbase.faker.validate

import io.agentcontextbase.faker.domain.TenantCoreDataset
import io.agentcontextbase.faker.pools.IdPools
import java.time.Instant
import java.time.temporal.ChronoUnit

data class ValidationReport(
    val ok: Boolean,
    val violations: List<String>,
    val counts: Map<String, Int>,
    val seed: Int,
    val profileName: String,
)

fun validate(dataset: TenantCoreDataset): ValidationReport {
    val violations = mutableListOf<String>()
    val counts = mapOf(
        "organizations" to dataset.organizations.size,
        "users" to dataset.users.size,
        "memberships" to dataset.memberships.size,
        "projects" to dataset.projects.size,
        "audit_events" to dataset.auditEvents.size,
        "api_keys" to dataset.apiKeys.size,
        "invitations" to dataset.invitations.size,
    )

    val orgs = mutableMapOf<String, io.agentcontextbase.faker.domain.Organization>()
    dataset.organizations.forEach { org ->
        if (orgs.put(org.id, org) != null) violations += "duplicate organizations.id: ${org.id}"
    }
    val users = mutableMapOf<String, io.agentcontextbase.faker.domain.User>()
    val seenEmails = mutableSetOf<String>()
    dataset.users.forEach { user ->
        users[user.id] = user
        if (!seenEmails.add(user.email.lowercase())) violations += "duplicate users.email: ${user.email}"
    }

    val membershipIds = mutableSetOf<String>()
    dataset.memberships.forEach { membership ->
        membershipIds += membership.id
        val org = orgs[membership.orgId]
        if (org == null) {
            violations += "membership missing org: ${membership.id}"
            return@forEach
        }
        if (users[membership.userId] == null) {
            violations += "membership missing user: ${membership.id}"
            return@forEach
        }
        if (Instant.parse(membership.joinedAt).isBefore(Instant.parse(org.createdAt))) {
            violations += "Rule A violated by membership ${membership.id}"
        }
        if (membership.invitedBy != null && users[membership.invitedBy] == null) {
            violations += "membership invited_by missing user: ${membership.id}"
        }
    }

    val projects = dataset.projects.associateBy { it.id }
    dataset.projects.forEach { project ->
        val org = orgs[project.orgId]
        if (org == null) {
            violations += "project missing org: ${project.id}"
            return@forEach
        }
        if (Instant.parse(project.createdAt).isBefore(Instant.parse(org.createdAt))) {
            violations += "Rule B violated by project ${project.id}"
        }
    }

    val graph = IdPools.graph(dataset)
    val apiKeyIds = mutableSetOf<String>()
    val keyPrefixes = mutableSetOf<String>()
    dataset.apiKeys.forEach { apiKey ->
        apiKeyIds += apiKey.id
        if (!graph.orgMemberMap[apiKey.orgId].orEmpty().contains(apiKey.createdBy)) {
            violations += "Rule G violated by api_key ${apiKey.id}"
        }
        if (!keyPrefixes.add(apiKey.keyPrefix)) {
            violations += "duplicate api_key.key_prefix: ${apiKey.keyPrefix}"
        }
        if (apiKey.lastUsedAt != null && Instant.parse(apiKey.lastUsedAt).isBefore(Instant.parse(apiKey.createdAt))) {
            violations += "api_key last_used_at before created_at: ${apiKey.id}"
        }
    }

    val invitationIds = mutableSetOf<String>()
    dataset.invitations.forEach { invitation ->
        invitationIds += invitation.id
        if (!graph.orgMemberMap[invitation.orgId].orEmpty().contains(invitation.invitedBy)) {
            violations += "Rule H violated by invitation ${invitation.id}"
        }
        val matchesMember = graph.orgMemberMap[invitation.orgId].orEmpty()
            .mapNotNull { graph.userEmailById[it] }
            .any { it.equals(invitation.invitedEmail, ignoreCase = true) }
        if (matchesMember) {
            violations += "Rule I violated by invitation ${invitation.id}"
        }
        val expiresAt = Instant.parse(invitation.expiresAt)
        if (!expiresAt.isAfter(IdPools.BASE_TIME)) {
            violations += "invitation expiry must be in the future: ${invitation.id}"
        }
        if (expiresAt.isAfter(IdPools.BASE_TIME.plus(30, ChronoUnit.DAYS))) {
            violations += "invitation expiry must be within 30 days: ${invitation.id}"
        }
        if (invitation.acceptedAt != null && Instant.parse(invitation.acceptedAt).isAfter(IdPools.BASE_TIME)) {
            violations += "invitation accepted_at must be in the past: ${invitation.id}"
        }
    }

    dataset.projects.forEach { project ->
        if (!graph.orgMemberMap[project.orgId].orEmpty().contains(project.createdBy)) {
            violations += "Rule C violated by project ${project.id}"
        }
    }

    dataset.auditEvents.forEach { event ->
        val project = projects[event.projectId]
        if (project == null) {
            violations += "audit_event missing project: ${event.id}"
            return@forEach
        }
        if (!graph.orgMemberMap[event.orgId].orEmpty().contains(event.userId)) {
            violations += "Rule D violated by audit_event ${event.id}"
        }
        if (project.orgId != event.orgId) {
            violations += "Rule E violated by audit_event ${event.id}"
        }
        if (Instant.parse(event.ts).isBefore(Instant.parse(project.createdAt))) {
            violations += "Rule F violated by audit_event ${event.id}"
        }
        val membership = graph.membershipByOrgUser["${event.orgId}:${event.userId}"]
        if (membership != null && Instant.parse(event.ts).isBefore(Instant.parse(membership.joinedAt))) {
            violations += "audit event before membership joined_at: ${event.id}"
        }
        when (event.resourceType) {
            "project" -> if (!projects.containsKey(event.resourceId)) violations += "audit event resource project missing: ${event.id}"
            "user" -> if (!users.containsKey(event.resourceId)) violations += "audit event resource user missing: ${event.id}"
            "membership" -> if (!membershipIds.contains(event.resourceId)) violations += "audit event resource membership missing: ${event.id}"
            "api_key" -> if (!apiKeyIds.contains(event.resourceId)) violations += "audit event resource api_key missing: ${event.id}"
            "invitation" -> if (!invitationIds.contains(event.resourceId)) violations += "audit event resource invitation missing: ${event.id}"
        }
    }

    return ValidationReport(
        ok = violations.isEmpty(),
        violations = violations,
        counts = counts,
        seed = dataset.seed,
        profileName = dataset.profileName,
    )
}
